File: config/default.go
package config

import (
	"github.com/ngerakines/preview/util"
	"os"
	"path/filepath"
)

func NewDefaultAppConfig() (AppConfig, error) {
	return buildDefaultConfig(defaultBasePath)
}

func NewDefaultAppConfigWithBaseDirectory(root string) (AppConfig, error) {
	return buildDefaultConfig(func(section string) string {
		cacheDirectory := filepath.Join(root, ".cache", section)
		os.MkdirAll(cacheDirectory, 00777)
		return cacheDirectory
	})
}

func buildDefaultConfig(basePathFunc basePath) (AppConfig, error) {
	config := `{
   "common": {
      "placeholderBasePath":"` + basePathFunc("placeholders") + `",
      "placeholderGroups": {
         "image":["jpg", "jpeg", "png", "gif"],
         "document":["pdf", "doc", "docx"]
      },
      "localAssetStoragePath":"` + basePathFunc("assets") + `",
      "nodeId":"E876F147E331"
   },
   "http":{
      "listen":":8080"
   },
   "storage":{
      "engine":"memory"
   },
   "documentRenderAgent":{
      "enabled":true,
      "count":16,
      "basePath":"` + basePathFunc("documentRenderAgentTmp") + `"
   },
   "imageMagickRenderAgent":{
      "enabled":true,
      "count":16,
      "supportedFileTypes":{
         "jpg":33554432,
         "jpeg":33554432,
         "png":33554432,
         "gif":33554432,
         "pdf":33554432
      }
   },
   "simpleApi":{
      "enabled":true,
      "edgeBaseUrl":"http://localhost:8080"
   },
   "assetApi":{
      "enabled":true
   },
   "uploader":{
      "engine":"local"
   },
   "downloader":{
      "basePath":"` + basePathFunc("cache") + `",
      "tramEnabled": false
   }
}`
	return NewUserAppConfig([]byte(config))
}

type basePath func(string) string

func defaultBasePath(section string) string {
	cacheDirectory := filepath.Join(util.Cwd(), ".cache", section)
	os.MkdirAll(cacheDirectory, 00777)
	return cacheDirectory
}

File: cookbooks/preview/attributes/default.rb

default[:preview][:platform] = 'amd64'
default[:preview][:version] = '0.1.0'
default[:preview][:install_type] = 'archive'
default[:preview][:package] = 'preview'
default[:preview][:archive_source] = "https://github.com/ngerakines/preview/releases/download/v#{node[:preview][:version]}/preview-#{node[:preview][:version]}-linux_#{node[:preview][:platform]}.zip"
# default[:preview][:archive_source] = "https://github.com/ngerakines/preview/releases/download/v0.1.0/preview-0.1.0-linux_amd64.zip"

default[:preview][:port] = 8080
default[:preview][:basePath] = "/home/preview/data/"
default[:preview][:config] = {}

default[:preview][:config][:common] = {}
default[:preview][:config][:common][:placeholderBasePath] = "#{node[:preview][:basePath]}placeholders"
default[:preview][:config][:common][:placeholderGroups] = {}
default[:preview][:config][:common][:placeholderGroups][:image] = ['jpg', 'jpeg', 'png', 'gif']
default[:preview][:config][:common][:placeholderGroups][:document] = ['pdf', 'doc', 'docx']
default[:preview][:config][:common][:placeholderGroups][:presentation] = ['ppt', 'pptx']
default[:preview][:config][:common][:localAssetStoragePath] = "#{node[:preview][:basePath]}assets"
default[:preview][:config][:common][:nodeId] = "E876F147E331"

default[:preview][:config][:http] = {}
default[:preview][:config][:http][:listen] = ":#{node[:preview][:port]}"

default[:preview][:config][:storage] = {}
default[:preview][:config][:storage][:engine] = "memory"

default[:preview][:config][:documentRenderAgent] = {}
default[:preview][:config][:documentRenderAgent][:enabled] = true
default[:preview][:config][:documentRenderAgent][:count] = 8
default[:preview][:config][:documentRenderAgent][:basePath] = "#{node[:preview][:basePath]}tmp/documentRenderAgent/"
default[:preview][:config][:documentRenderAgent][:supportedFileTypes] = {}
default[:preview][:config][:documentRenderAgent][:supportedFileTypes][:doc] = 33554432
default[:preview][:config][:documentRenderAgent][:supportedFileTypes][:docx] = 33554432
default[:preview][:config][:documentRenderAgent][:supportedFileTypes][:ppt] = 33554432
default[:preview][:config][:documentRenderAgent][:supportedFileTypes][:pptx] = 33554432

default[:preview][:config][:imageMagickRenderAgent] = {}
default[:preview][:config][:imageMagickRenderAgent][:enabled] = true
default[:preview][:config][:imageMagickRenderAgent][:count] = 8
default[:preview][:config][:imageMagickRenderAgent][:supportedFileTypes] = {}
default[:preview][:config][:imageMagickRenderAgent][:supportedFileTypes][:jpg] = 33554432
default[:preview][:config][:imageMagickRenderAgent][:supportedFileTypes][:jpeg] = 33554432
default[:preview][:config][:imageMagickRenderAgent][:supportedFileTypes][:png] = 33554432
default[:preview][:config][:imageMagickRenderAgent][:supportedFileTypes][:gif] = 33554432
default[:preview][:config][:imageMagickRenderAgent][:supportedFileTypes][:pdf] = 33554432

default[:preview][:config][:simpleApi] = {}
default[:preview][:config][:simpleApi][:enabled] = true
default[:preview][:config][:simpleApi][:baseUrl] = "/api"
default[:preview][:config][:simpleApi][:edgeBaseUrl] = "http://localhost:#{node[:preview][:port]}"

default[:preview][:config][:assetApi] = {}
default[:preview][:config][:assetApi][:enabled] = true

default[:preview][:config][:uploader] = {}
default[:preview][:config][:uploader][:engine] = "local"

default[:preview][:config][:downloader] = {}
default[:preview][:config][:downloader][:basePath] = "#{node[:preview][:basePath]}tmp/downloads/"
default[:preview][:config][:downloader][:tramEnabled] = false
